akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessState.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.typed.internal
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.cluster.ddata.LWWRegister
+import akka.cluster.ddata.LWWRegisterKey
+import akka.cluster.ddata.typed.scaladsl.DistributedData
+import akka.cluster.ddata.typed.scaladsl.Replicator
+import akka.cluster.sharding.typed.ShardedDaemonProcessContext
+
+import java.time.Instant
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class ShardedDaemonProcessState(
+    revision: Long,
+    numberOfProcesses: Int,
+    completed: Boolean,
+    started: Instant)
+
+/**
+ * INTERNAL API
+ */
+private[akka] object ShardedDaemonProcessState {
+
+  val startRevision = 0L
+
+  type Register = LWWRegister[ShardedDaemonProcessState]
+
+  def ddataKey(name: String): LWWRegisterKey[ShardedDaemonProcessState] =
+    LWWRegisterKey[ShardedDaemonProcessState](name)
+
+  def verifyRevisionBeforeStarting[T](
+      behaviorFactory: ShardedDaemonProcessContext => Behavior[T]): ShardedDaemonProcessContext => Behavior[T] = {
+    sdpContext =>
+      Behaviors.setup { context =>
+        val revision = sdpContext.revision
+        val key = ddataKey(sdpContext.name)
+
+        context.log.debug2(
+          "{}: Deferred start of worker to verify its revision [{}] is the latest",
+          sdpContext.name,
+          revision)
+
+        // we can't anyway turn reply into T so no need for the usual adapter
+        val distributedData = DistributedData(context.system)
+        distributedData.replicator ! Replicator.Get(key, Replicator.ReadLocal, context.self.unsafeUpcast)
+        Behaviors.receiveMessagePartial {
+          case reply @ Replicator.GetSuccess(`key`) =>
+            val state = reply.get(key).value
+            if (state.revision == revision) {
+              context.log.debug2("{}: Starting worker, revision [{}] is the latest", sdpContext.name, revision)
+              behaviorFactory(sdpContext).unsafeCast
+            } else {
+              context.log.warnN(
+                "{}: Tried to start an old revision of worker ([{}] but latest revision is [{}], started at {})",
+                sdpContext.name,
+                sdpContext.revision,
+                state.revision,
+                state.started)
+              Behaviors.stopped // FIXME do we need to passivate or crash rather than stop?
```

Comment:
no need for passivation afaics, there are no important incoming messages in mailbox